server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -322,7 +325,13 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
   {
     try {
       log.info("Loading segment %s", segment.getIdentifier());
-      if (loadSegment(segment, callback)) {
+      if (segmentsToDelete.contains(segment)) {
+        synchronized (lock) {
+          segmentsToDelete.remove(segment);
+        }
+      }
```

Comment:
can we move both the contains and remove inside the synchronized block ? Keeping it outside should be theoretically fine since the set is concurrent, just makes the understanding the synchronization a bit harder.